### PR TITLE
Revert PR #7241 to fix vault renewal

### PR DESCRIPTION
**Pull request by poornas**
and check if non-null secret was returned by vault login

- Current implementation is spawning renewer goroutines
without waiting for the lease duration to end.

The issue being fixed by #7241  1 needs to only check if secret returned upon login is nil, as might be the case of a HTTP 404.

## Regression
 Yes

## How Has This Been Tested?

Manually with vault as KMS settings enabled
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by badscooter23** [comment]
It makes sense to revert this fix as it was causing a different problem, but just want to clarify that we are still having the problem the fix was originally intended to fix and it is causing downtime. So we need a stable fix t the originl problem.

**Comment by harshavardhana** [comment]
> It makes sense to revert this fix as it was causing a different problem, but just want to clarify that we are still having the problem the fix was originally intended to fix and it is causing downtime. So we need a stable fix t the originl problem.

This fix also fixes the original problem the #7241 was supposed to fix @badscooter23

**Comment by badscooter23** [comment]
> > It makes sense to revert this fix as it was causing a different problem, but just want to clarify that we are still having the problem the fix was originally intended to fix and it is causing downtime. So we need a stable fix t the originl problem.
> 
> This fix also fixes the original problem the #7241 was supposed to fix @badscooter23

OK great! :) Thanks!

File: cmd/crypto/vault.go [hunk 1]
```diff
@@ -130,79 +130,70 @@ func NewVault(config VaultConfig) (KMS, error) {
...
+
+// renewer calls vault client's renewer that automatically
+// renews secret periodically
+func (v *vaultService) renewer(secret *vault.Secret) {
 	renewer, err := v.client.NewRenewer(&vault.RenewerInput{
 		Secret: secret,
 	})
 	if err != nil {
-		logger.CriticalIf(context.Background(), fmt.Errorf("crypto: failed to create hashicorp vault renewer: %s", err))
+		logger.FatalIf(err, "crypto: hashicorp vault token renewer could not be started")
```
Comment (harshavardhana) [inline cmd/crypto/vault.go]: Let's keep this logger.CriticalIf() - FatalIf is not a good idea here @poornas

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |

File: cmd/crypto/vault.go [hunk 2]
```diff
@@ -130,79 +130,70 @@ func NewVault(config VaultConfig) (KMS, error) {
...
 	go renewer.Renew()
 	defer renewer.Stop()
 
 	for {
 		select {
 		case err := <-renewer.DoneCh():
 			if err != nil {
-				return nil, err
+				v.reauthenticate()
+				renewer.Stop()
```
Comment (harshavardhana) [inline cmd/crypto/vault.go]: We already do `defer` so why need to do explicit `renewer.Stop()` ?

File: cmd/crypto/vault.go [hunk 3]
```diff
@@ -130,79 +130,70 @@ func NewVault(config VaultConfig) (KMS, error) {
...
 			}
-		case renew := <-renewer.RenewCh():
-			if renew.Secret == nil || renew.Secret.Auth == nil {
-				return nil, ErrKMSAuthLogin
-			}
-			return renew.Secret, nil
+
+			// Renewal is now over
+		case renewal := <-renewer.RenewCh():
+			v.leaseDuration = time.Duration(renewal.Secret.Auth.LeaseDuration)
```
Comment (harshavardhana) [inline cmd/crypto/vault.go]: This variable update is not needed unless we use the `renewal.Secret.Auth.Token`  - also we can remove the leaseDuration variable, i don't see where its used.

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |

File: cmd/crypto/vault.go (hunk 1 above)
Comment (poornas) [reply cmd/crypto/vault.go]: fixed.

File: cmd/crypto/vault.go (hunk 2 above)
Comment (poornas) [reply cmd/crypto/vault.go]: unnecessary, so removed it.

File: cmd/crypto/vault.go [hunk 4]
```diff
@@ -130,79 +130,68 @@ func NewVault(config VaultConfig) (KMS, error) {
...
 			}
-		case renew := <-renewer.RenewCh():
-			if renew.Secret == nil || renew.Secret.Auth == nil {
-				return nil, ErrKMSAuthLogin
-			}
-			return renew.Secret, nil
+
+			// Renewal is now over
+		case <-renewer.RenewCh():
+			return
```
Comment (harshavardhana) [inline cmd/crypto/vault.go]: We shouldn't return here @poornas we need renewer to be running always .. i.e RenewCh() will return the renewed token. We also don't need a new `v.tokenRenewer` its not used by anyone..

Comment (harshavardhana) [reply cmd/crypto/vault.go]: The renewed token returned by RenewCh should be used again to set the AuthToken() .and also used with the NewRenewer()

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: cmd/crypto/vault.go (hunk 4 above)
Comment (poornas) [reply cmd/crypto/vault.go]: when we reauthenticate, we start a new renewer

Comment (harshavardhana) [reply cmd/crypto/vault.go]: We shouldn't - that is not the flow from vault SDK we are incurring costs which are not needed.. @poornas

Comment (aead) [reply cmd/crypto/vault.go]: https://godoc.org/github.com/hashicorp/vault/api#Renewer

Comment (aead) [reply cmd/crypto/vault.go]: Also notice that the renewed secret may be `nil` - See: https://github.com/minio/minio/pull/7259/files#diff-bc9558a54da677c8f7933fee0f0c42f5L152

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :x: [more...](#edb9b564-d598-4ba6-b958-9a9f57f5b45e) |
| mint-dist-xl.sh | :x: [more...](#b25705e0-0e37-477c-9250-823905d5470a) |
| mint-gateway-nas.sh | :x: [more...](#cfee9e6a-5695-4f55-ba23-5ee420aa7b85) |
<a name="edb9b564-d598-4ba6-b958-9a9f57f5b45e"></a>
### 7259-1bca31c/mint-compression-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.60:31435
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 47edc63684b4:/mint/log /tmp/mint-logs'

(1/13) Running aws-sdk-go tests ... done in 0 seconds
(2/13) Running aws-sdk-java tests ... done in 0 seconds
(3/13) Running aws-sdk-php tests ... done in 42 seconds
(4/13) Running aws-sdk-ruby tests ... done in 2 seconds
(5/13) Running awscli tests ... done in 1 minutes and 44 seconds
(6/13) Running mc tests ... done in 19 seconds
(7/13) Running minio-dotnet tests ... done in 54 seconds
(8/13) Running minio-go tests ... done in 20 seconds
(9/13) Running minio-java tests ... FAILED in 20 seconds
{
  "name": "minio-java",
  "function": "putObject(String bucketName, String objectName, InputStream body, long size, String contentType)",
  "args": "contentType: application/javascript",
  "duration": 16488,
  "status": "FAIL",
  "error": "java.net.SocketException: Connection reset >>> [java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:115), java.net.SocketOutputStream.write(SocketOutputStream.java:155), okio.Okio$1.write(Okio.java:78), okio.AsyncTimeout$1.write(AsyncTimeout.java:179), okio.RealBufferedSink.emitCompleteSegments(RealBufferedSink.java:171), okio.RealBufferedSink.write(RealBufferedSink.java:41), okhttp3.internal.http1.Http1Codec$FixedLengthSink.write(Http1Codec.java:286), okio.RealBufferedSink.emitCompleteSegments(RealBufferedSink.java:171), okio.RealBufferedSink.write(RealBufferedSink.java:109), io.minio.HttpRequestBody.writeTo(HttpRequestBody.java:74), okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:62), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92), okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:45), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67), okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67), okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92), okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:120), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67), okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:185), okhttp3.RealCall.execute(RealCall.java:69), io.minio.MinioClient.executeReq(MinioClient.java:1097), io.minio.MinioClient.execute(MinioClient.java:1056), io.minio.MinioClient.executePut(MinioClient.java:1399), io.minio.MinioClient.executePut(MinioClient.java:1421), io.minio.MinioClient.putObject(MinioClient.java:3994), io.minio.MinioClient.putObject(MinioClient.java:4086), io.minio.MinioClient.putObject(MinioClient.java:3960), FunctionalTest.testPutObjectUnknownStreamSize(FunctionalTest.java:556), FunctionalTest.putObject_test8(FunctionalTest.java:582), FunctionalTest.runTests(FunctionalTest.java:2998), FunctionalTest.main(FunctionalTest.java:3154)]"
}

Executed 8 out of 13 tests successfully.
```
<a name="b25705e0-0e37-477c-9250-823905d5470a"></a>
### 7259-1bca31c/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.60:31977
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 69cc734ac7e8:/mint/log /tmp/mint-logs'

(1/13) Running aws-sdk-go tests ... done in 0 seconds
(2/13) Running aws-sdk-java tests ... done in 1 seconds
(3/13) Running aws-sdk-php tests ... done in 45 seconds
(4/13) Running aws-sdk-ruby tests ... done in 4 seconds
(5/13) Running awscli tests ... done in 35 minutes and 13 seconds
(6/13) Running mc tests ... FAILED in 1 minutes and 31 seconds
{
  "name": "mc",
  "duration": "60197",
  "function": "test_put_object",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-21833 --quiet --no-color rm myminio/mc-test-bucket-30602/mc-test-object-1613  >>> \nmc: <ERROR> Failed to remove `myminio/mc-test-bucket-30602/mc-test-object-1613`. Get http://72.28.97.60:31977/mc-test-bucket-30602/?location=: dial tcp 72.28.97.60:31977: i/o timeout"
}

Executed 5 out of 13 tests successfully.
```
<a name="cfee9e6a-5695-4f55-ba23-5ee420aa7b85"></a>
### 7259-1bca31c/mint-gateway-nas.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.54:31676
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 50f4f2441f56:/mint/log /tmp/mint-logs'

(1/13) Running aws-sdk-go tests ... done in 0 seconds
(2/13) Running aws-sdk-java tests ... done in 1 seconds
(3/13) Running aws-sdk-php tests ... done in 41 seconds
(4/13) Running aws-sdk-ruby tests ... done in 2 seconds
(5/13) Running awscli tests ... done in 48 seconds
(6/13) Running mc tests ... done in 9 seconds
(7/13) Running minio-dotnet tests ... done in 47 seconds
(8/13) Running minio-go tests ... done in 23 seconds
(9/13) Running minio-java tests ... FAILED in 8 seconds
{
  "name": "minio-java",
  "function": "putObject(String bucketName, String objectName, InputStream body, long size, String contentType)",
  "args": "contentType: application/javascript",
  "duration": 4125,
  "status": "FAIL",
  "error": "java.net.SocketException: Connection reset >>> [java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:115), java.net.SocketOutputStream.write(SocketOutputStream.java:155), okio.Okio$1.write(Okio.java:78), okio.AsyncTimeout$1.write(AsyncTimeout.java:179), okio.RealBufferedSink.emitCompleteSegments(RealBufferedSink.java:171), okio.RealBufferedSink.write(RealBufferedSink.java:41), okhttp3.internal.http1.Http1Codec$FixedLengthSink.write(Http1Codec.java:286), okio.RealBufferedSink.emitCompleteSegments(RealBufferedSink.java:171), okio.RealBufferedSink.write(RealBufferedSink.java:109), io.minio.HttpRequestBody.writeTo(HttpRequestBody.java:74), okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:62), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92), okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:45), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67), okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67), okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92), okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:120), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67), okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:185), okhttp3.RealCall.execute(RealCall.java:69), io.minio.MinioClient.executeReq(MinioClient.java:1097), io.minio.MinioClient.execute(MinioClient.java:1056), io.minio.MinioClient.executePut(MinioClient.java:1399), io.minio.MinioClient.executePut(MinioClient.java:1421), io.minio.MinioClient.putObject(MinioClient.java:3994), io.minio.MinioClient.putObject(MinioClient.java:4086), io.minio.MinioClient.putObject(MinioClient.java:3960), FunctionalTest.testPutObjectUnknownStreamSize(FunctionalTest.java:556), FunctionalTest.putObject_test8(FunctionalTest.java:582), FunctionalTest.runTests(FunctionalTest.java:2998), FunctionalTest.main(FunctionalTest.java:3154)]"
}

Executed 8 out of 13 tests successfully.
```

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :x: [more...](#fbbad248-c2a8-4fcc-9d11-851978dc9e6b) |
<a name="fbbad248-c2a8-4fcc-9d11-851978dc9e6b"></a>
### 7259-8515d04/mint-compression-fs.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.56:30912
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 668120df1bee:/mint/log /tmp/mint-logs'

(1/13) Running aws-sdk-go tests ... done in 0 seconds
(2/13) Running aws-sdk-java tests ... done in 1 seconds
(3/13) Running aws-sdk-php tests ... done in 41 seconds
(4/13) Running aws-sdk-ruby tests ... done in 1 seconds
(5/13) Running awscli tests ... done in 46 seconds
(6/13) Running mc tests ... done in 11 seconds
(7/13) Running minio-dotnet tests ... done in 48 seconds
(8/13) Running minio-go tests ... done in 25 seconds
(9/13) Running minio-java tests ... FAILED in 19 seconds
{
  "name": "minio-java",
  "function": "putObject(String bucketName, String objectName, InputStream body, long size, String contentType)",
  "args": "contentType: application/javascript",
  "duration": 13928,
  "status": "FAIL",
  "error": "java.net.SocketException: Connection reset >>> [java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:115), java.net.SocketOutputStream.write(SocketOutputStream.java:155), okio.Okio$1.write(Okio.java:78), okio.AsyncTimeout$1.write(AsyncTimeout.java:179), okio.RealBufferedSink.emitCompleteSegments(RealBufferedSink.java:171), okio.RealBufferedSink.write(RealBufferedSink.java:41), okhttp3.internal.http1.Http1Codec$FixedLengthSink.write(Http1Codec.java:286), okio.RealBufferedSink.emitCompleteSegments(RealBufferedSink.java:171), okio.RealBufferedSink.write(RealBufferedSink.java:109), io.minio.HttpRequestBody.writeTo(HttpRequestBody.java:74), okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:62), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92), okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:45), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67), okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67), okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92), okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:120), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67), okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:185), okhttp3.RealCall.execute(RealCall.java:69), io.minio.MinioClient.executeReq(MinioClient.java:1097), io.minio.MinioClient.execute(MinioClient.java:1056), io.minio.MinioClient.executePut(MinioClient.java:1399), io.minio.MinioClient.executePut(MinioClient.java:1421), io.minio.MinioClient.putObject(MinioClient.java:3994), io.minio.MinioClient.putObject(MinioClient.java:4086), io.minio.MinioClient.putObject(MinioClient.java:3960), FunctionalTest.testPutObjectUnknownStreamSize(FunctionalTest.java:556), FunctionalTest.putObject_test8(FunctionalTest.java:582), FunctionalTest.runTests(FunctionalTest.java:2998), FunctionalTest.main(FunctionalTest.java:3154)]"
}

Executed 8 out of 13 tests successfully.
```

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |

**Comment by harshavardhana** [comment]
The DB growth issue is fixed after this PR - here is the numbers of every second with the master

```
ETCDCTL_API=3 etcdctl --endpoints http://localhost:2379 endpoint status -w table
+-----------------------+------------------+---------+---------+-----------+-----------+------------+
|       ENDPOINT        |        ID        | VERSION | DB SIZE | IS LEADER | RAFT TERM | RAFT INDEX |
+-----------------------+------------------+---------+---------+-----------+-----------+------------+
| http://localhost:2379 | 8e9e05c52164694d |   3.3.9 |   18 MB |      true |         2 |       8704 |
+-----------------------+------------------+---------+---------+-----------+-----------+------------+
```
```
 ETCDCTL_API=3 etcdctl --endpoints http://localhost:2379 endpoint status -w table
+-----------------------+------------------+---------+---------+-----------+-----------+------------+
|       ENDPOINT        |        ID        | VERSION | DB SIZE | IS LEADER | RAFT TERM | RAFT INDEX |
+-----------------------+------------------+---------+---------+-----------+-----------+------------+
| http://localhost:2379 | 8e9e05c52164694d |   3.3.9 |   20 MB |      true |         2 |       9562 |
+-----------------------+------------------+---------+---------+-----------+-----------+------------+
```

```
 ETCDCTL_API=3 etcdctl --endpoints http://localhost:2379 endpoint status -w table
+-----------------------+------------------+---------+---------+-----------+-----------+------------+
|       ENDPOINT        |        ID        | VERSION | DB SIZE | IS LEADER | RAFT TERM | RAFT INDEX |
+-----------------------+------------------+---------+---------+-----------+-----------+------------+
| http://localhost:2379 | 8e9e05c52164694d |   3.3.9 |   22 MB |      true |         2 |      10837 |
+-----------------------+------------------+---------+---------+-----------+-----------+------------+
```

As you can see consistent growth - but with this PR no growth at all. Until the lease duration which is quite minimal. The PR is good with the etcd table growth point of view.

File: cmd/crypto/vault.go [hunk 5]
```diff
@@ -123,86 +121,69 @@ func NewVault(config VaultConfig) (KMS, error) {
...
+				time.Sleep(retryDelay)
+				continue
 			}
-			return renew.Secret, nil
+			if ok, err := s.TokenIsRenewable(); !ok || err != nil {
+				v.secret = nil
+				continue
+			}
+			v.secret = s
+			nextRenew := s.Auth.LeaseDuration / 2
```
Comment (aead) [inline cmd/crypto/vault.go]: Is it guaranteed that the `LeaseDuration` in https://godoc.org/github.com/hashicorp/vault/api#SecretAuth will be in sec and not e.g. minutes?

Comment (poornas) [reply cmd/crypto/vault.go]: @aead, internally their renewer assumes the lease duration is in seconds - so it should be safe to assume the same

Comment (harshavardhana) [reply cmd/crypto/vault.go]: Yes, its guaranteed @aead - that's the logic even renewAuth() code has. https://www.vaultproject.io/api/system/leases.html#renew-lease

**Review by harshavardhana — DISMISSED** [review]

File: cmd/crypto/vault.go (hunk 5 above)
Comment (aead) [reply cmd/crypto/vault.go]: Okay

**Review by aead — DISMISSED** [review]

**Comment by poornas** [comment]
@aead and @harshavardhana , hold the approvals - think there would be an issue if lease duration is less than a minute, will update with a fix

**Comment by poornas** [comment]
> @aead and @harshavardhana , hold the approvals - think there would be an issue if lease duration is less than a minute, will update with a fix

can you both please  review again - it was wrong to assume a retryDelay of 1 minute if the original token renewal periodicity was set to something lower. So setting retryDelay from lease duration returned upon login,

**Review by harshavardhana — DISMISSED** [review]
LGTM and tested again.

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |

File: cmd/crypto/vault.go [hunk 6]
```diff
@@ -123,86 +122,79 @@ func NewVault(config VaultConfig) (KMS, error) {
...
 
-// authenticate tries to authenticate the minio server
-// to the Vault KMS and starts a background job to renew
-// the login.
-func (v *vaultService) authenticate() error {
-	secret, err := v.login()
+	tokenID, err = secret.TokenID()
 	if err != nil {
-		return err
+		return
```
Comment (aead) [inline cmd/crypto/vault.go:189]: We should return `ErrKMSAuthLogin` to the S3 client

File: cmd/crypto/vault.go [hunk 7]
```diff
@@ -123,86 +122,79 @@ func NewVault(config VaultConfig) (KMS, error) {
...
-			}
-			secret = newSecret // Now newSecret contains a valid, non-nil *vault.Secret
-			v.client.SetToken(secret.Auth.ClientToken)
-			v.leaseDuration = time.Duration(secret.Auth.LeaseDuration)
-		}
-	}(secret)
-	return nil
+	ttl, err = secret.TokenTTL()
+	if err != nil {
+		return
```
Comment (aead) [inline cmd/crypto/vault.go:194]: We should return `ErrKMSAuthLogin` to the S3 client.

File: cmd/crypto/vault.go (hunk 6 above)
Comment (poornas) [reply cmd/crypto/vault.go:189]: done @aead

File: cmd/crypto/vault.go (hunk 7 above)
Comment (poornas) [reply cmd/crypto/vault.go:194]: fixed

**Review by aead — APPROVED** [review]

**Review by harshavardhana — APPROVED** [review]

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-xl.sh | :x: [more...](#ea352c99-bf64-4760-83f8-c14e285a6f75) |
<a name="ea352c99-bf64-4760-83f8-c14e285a6f75"></a>
### 7259-6801230/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.61:31226
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp de1280eaa4ac:/mint/log /tmp/mint-logs'

(1/13) Running aws-sdk-go tests ... done in 0 seconds
(2/13) Running aws-sdk-java tests ... done in 1 seconds
(3/13) Running aws-sdk-php tests ... done in 41 seconds
(4/13) Running aws-sdk-ruby tests ... done in 2 seconds
(5/13) Running awscli tests ... done in 1 minutes and 43 seconds
(6/13) Running mc tests ... done in 12 seconds
(7/13) Running minio-dotnet tests ... done in 47 seconds
(8/13) Running minio-go tests ... done in 26 seconds
(9/13) Running minio-java tests ... done in 51 seconds
(10/13) Running minio-js tests ... done in 36 seconds
(11/13) Running minio-py tests ... done in 1 minutes and 44 seconds
(12/13) Running s3cmd tests ... FAILED in 6 seconds
{
  "name": "s3cmd",
  "duration": "2132",
  "function": "test_get_object_multipart",
  "status": "FAIL",
  "error": "\n!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n    An unexpected error has occurred.\n  Please try reproducing the error using\n  the latest s3cmd code from the git master\n  branch found at:\n    https://github.com/s3tools/s3cmd\n  and have a look at the known issues list:\n    https://github.com/s3tools/s3cmd/wiki/Common-known-issues-and-their-solutions\n  If the error persists, please report the\n  following lines (removing any private\n  info as necessary) to:\n   s3tools-bugs@lists.sourceforge.net\n\n\n!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n\nInvoked as: /usr/local/bin/s3cmd --config /tmp/.s3cmd-31622/s3cfg get s3://s3cmd-test-bucket-16870/s3cmd-test-object-28517 s3cmd-test-object-28517.downloaded\nProblem: <class 'ConnectionResetError: [Errno 104] Connection reset by peer\nS3cmd:   2.0.2\npython:   3.5.2 (default, Nov 12 2018, 13:43:14) \n[GCC 5.4.0 20160609]\nenvironment LANG=C.UTF-8\n\nTraceback (most recent call last):\n  File \"/usr/local/bin/s3cmd\", line 3092, in <module>\n    rc = main()\n  File \"/usr/local/bin/s3cmd\", line 3001, in main\n    rc = cmd_func(args)\n  File \"/usr/local/bin/s3cmd\", line 590, in cmd_object_get\n    response = s3.object_get(uri, dst_stream, destination, start_position = start_position, extra_label = seq_label)\n  File \"/usr/local/lib/python3.5/dist-packages/S3/S3.py\", line 716, in object_get\n    response = self.recv_file(request, stream, labels, start_position)\n  File \"/usr/local/lib/python3.5/dist-packages/S3/S3.py\", line 1742, in recv_file\n    data = http_response.read(this_chunk)\n  File \"/usr/lib/python3.5/http/client.py\", line 448, in read\n    n = self.readinto(b)\n  File \"/usr/lib/python3.5/http/client.py\", line 488, in readinto\n    n = self.fp.readinto(b)\n  File \"/usr/lib/python3.5/socket.py\", line 575, in readinto\n    return self._sock.recv_into(b)\nConnectionResetError: [Errno 104] Connection reset by peer\n\n!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n    An unexpected error has occurred.\n  Please try reproducing the error using\n  the latest s3cmd code from the git master\n  branch found at:\n    https://github.com/s3tools/s3cmd\n  and have a look at the known issues list:\n    https://github.com/s3tools/s3cmd/wiki/Common-known-issues-and-their-solutions\n  If the error persists, please report the\n  above lines (removing any private\n  info as necessary) to:\n   s3tools-bugs@lists.sourceforge.net\n!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!"
}

Executed 11 out of 13 tests successfully.
```

**Pull request merged**